### PR TITLE
Make Trajectory and Idx 5D

### PR DESCRIPTION
**Pull request by fzimmermann89**
Would be great if somebody could lend another pair of eyes if I got all tests.

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #683
* #675
* #674
* #673
* #672
* #671
* __->__ #660

File: tests/operators/test_non_uniform_fast_fourier_op.py [hunk 1]
```diff
@@ -180,13 +180,13 @@ def test_non_uniform_fast_fourier_op_error_matrix():
...
+            traj=KTrajectory.from_tensor(torch.ones((3, 1, 1, 1, 1, 1))),
         )
 
     with pytest.raises(ValueError, match='encoding_matrix should have'):
         NonUniformFastFourierOp(
             direction=(-2, -1),
             recon_matrix=[1, 1],
             encoding_matrix=[1],
-            traj=KTrajectory.from_tensor(torch.randn((3, 1, 1, 1, 1))),
+            traj=KTrajectory.from_tensor(torch.ones((3, 1, 1, 1, 1, 1))),
```
Comment (koflera) [inline tests/operators/test_non_uniform_fast_fourier_op.py:188]: why ones instead of randn?

File: tests/operators/test_non_uniform_fast_fourier_op.py [hunk 2]
```diff
@@ -180,13 +180,13 @@ def test_non_uniform_fast_fourier_op_error_matrix():
             direction=(-2, -1),
             recon_matrix=[1, 1, 1],
             encoding_matrix=[1, 1],
-            traj=KTrajectory.from_tensor(torch.randn((3, 1, 1, 1, 1))),
+            traj=KTrajectory.from_tensor(torch.ones((3, 1, 1, 1, 1, 1))),
```
Comment (koflera) [inline tests/operators/test_non_uniform_fast_fourier_op.py:180]: why ones instead of randn?

File: tests/operators/test_non_uniform_fast_fourier_op.py [hunk 3]
```diff
@@ -169,7 +169,7 @@ def test_non_uniform_fast_fourier_op_error_directions():
             direction=(-2, -1, 'x'),
             recon_matrix=[1, 1],
             encoding_matrix=[1, 1],
-            traj=KTrajectory.from_tensor(torch.randn((3, 1, 1, 1, 1))),
+            traj=KTrajectory.from_tensor(torch.ones((3, 1, 1, 1, 1, 1))),
```
Comment (koflera) [inline tests/operators/test_non_uniform_fast_fourier_op.py:169]: why ones instead of randn?

File: src/mrpro/data/KTrajectory.py [hunk 4]
```diff
@@ -71,8 +71,8 @@ def as_any_float(tensor: torch.Tensor) -> torch.Tensor:
         except RuntimeError:
             raise ValueError('The k-space trajectory dimensions must be broadcastable.') from None
 
-        if len(shape) < 4:
-            raise ValueError('The k-space trajectory tensors should each have at least 4 dimensions.')
+        if len(shape) < 5:
+            raise ValueError('The k-space trajectory tensors should each have at least 5 dimensions.')
```
Comment (koflera) [inline src/mrpro/data/KTrajectory.py:75]: in the docstring of "KTrajectory", shouldn't `(*other, k2, k1, k0)` be replaced by `(*other, 1, k2, k1, k0)`?

**Comment by koflera** [comment]
In 
https://github.com/PTB-MR/mrpro/blob/20b1d10397e15a1c1e2fc54353cab7e08372be00/tests/data/test_trajectory.py#L153

`im_shape, k_shape` are not used. same for 

https://github.com/PTB-MR/mrpro/blob/20b1d10397e15a1c1e2fc54353cab7e08372be00/tests/data/test_trajectory.py#L184

**Comment by koflera** [comment]
`k_shape `not used in 
https://github.com/PTB-MR/mrpro/blob/20b1d10397e15a1c1e2fc54353cab7e08372be00/tests/operators/test_non_uniform_fast_fourier_op.py#L15

**Comment by fzimmermann89** [comment]
The CommonMrTrajectories  decorator adds the arguments. 
So if any function needs the image or k shape (which some do) we need to have these parameters in all functions.

We could if they are unused, add an underscore to the name. But is it worth it?

**Comment by fzimmermann89** [comment]
> `k_shape `not used in 
> https://github.com/PTB-MR/mrpro/blob/20b1d10397e15a1c1e2fc54353cab7e08372be00/tests/operators/test_non_uniform_fast_fourier_op.py#L15
> 

But here it should be removed, yes

**Comment by koflera** [comment]
missing coil dimension in the following?

https://github.com/PTB-MR/mrpro/blob/20b1d10397e15a1c1e2fc54353cab7e08372be00/src/mrpro/data/traj_calculators/KTrajectorySunflowerGoldenRpe.py#L85

File: src/mrpro/operators/CartesianSamplingOp.py [hunk 5]
```diff
@@ -50,23 +50,29 @@ def __init__(self, encoding_matrix: SpatialDimension[int], traj: KTrajectory) ->
...
             sorted_grid_shape.z = ktraj_tensor.shape[-3]
-            kz_idx = repeat(torch.arange(ktraj_tensor.shape[-3]), 'k2->other k2 k1 k0', other=1, k1=1, k0=1)
+            kz_idx = repeat(
+                torch.arange(ktraj_tensor.shape[-3]), 'k2->other coils k2 k1 k0', other=1, coils=1, k1=1, k0=1
+            )
 
         # 1D indices into a flattened tensor.
         kidx = kz_idx * sorted_grid_shape.y * sorted_grid_shape.x + ky_idx * sorted_grid_shape.x + kx_idx
-        kidx = rearrange(kidx, '... kz ky kx -> ... 1 (kz ky kx)')
+        kidx = rearrange(kidx, '... 1 kz ky kx -> ... 1 (kz ky kx)')
```
Comment (koflera) [inline src/mrpro/operators/CartesianSamplingOp.py]: maybe more obvious by 
```suggestion
        kidx = rearrange(kidx, '... coils kz ky kx -> ... coils (kz ky kx)', coils=1)
```

File: src/mrpro/operators/CartesianSamplingOp.py [hunk 6]
```diff
@@ -81,7 +87,7 @@ def __init__(self, encoding_matrix: SpatialDimension[int], traj: KTrajectory) ->
                 stacklevel=2,
             )
 
-            inside_encoding_matrix = rearrange(inside_encoding_matrix, '... kz ky kx -> ... 1 (kz ky kx)')
+            inside_encoding_matrix = rearrange(inside_encoding_matrix, '... 1 kz ky kx -> ... 1 (kz ky kx)')
```
Comment (koflera) [inline src/mrpro/operators/CartesianSamplingOp.py]: here as well?
```suggestion
            inside_encoding_matrix = rearrange(inside_encoding_matrix, '... coils kz ky kx -> ... coils (kz ky kx)', coils=1)
```

**Comment by koflera** [comment]
replace 1 by coils=1 in ?

https://github.com/PTB-MR/mrpro/blob/20b1d10397e15a1c1e2fc54353cab7e08372be00/src/mrpro/operators/CartesianSamplingOp.py#L90

File: tests/data/conftest.py [hunk 7]
```diff
@@ -23,7 +23,8 @@ def generate(n_k2: int, n_k1: int, n_k0: int, jitter: float) -> tuple[torch.Tens
             kx = kx + generator.float32_tensor((n_k2, n_k1, n_k0), high=jitter)
             ky = ky + generator.float32_tensor((n_k2, n_k1, n_k0), high=jitter)
             kz = kz + generator.float32_tensor((n_k2, n_k1, n_k0), high=jitter)
-        return kz.unsqueeze(0), ky.unsqueeze(0), kx.unsqueeze(0)
+        kz, ky, kx = (unsqueeze_left(x, 2) for x in (kz, ky, kx))
```
Comment (koflera) [inline tests/data/conftest.py:33]: maybe add a comment of what the resulting shape of `kz, ky `and `kx `will be

File: tests/data/test_kdata.py [hunk 8]
```diff
@@ -281,48 +267,21 @@ def test_KData_rearrange_k2_k1_into_k1(consistently_shaped_kdata):
 )
 def test_KData_split_k1_into_other(consistently_shaped_kdata, monkeypatch, n_other_split, other_label):
```
Comment (koflera) [inline tests/data/test_kdata.py]: monkeypatch not used?

**Review by koflera — CHANGES_REQUESTED** [review]
some very small comments, I was not able to find anything obviously wrong

File: tests/operators/test_non_uniform_fast_fourier_op.py (hunk 1 above)
Comment (fzimmermann89) [reply tests/operators/test_non_uniform_fast_fourier_op.py:188]: we dont want to use torch.randn without seed ever.
here, it really does not matter. it can be any value and does not have to be random for the test.

File: src/mrpro/data/KTrajectory.py (hunk 4 above)
Comment (fzimmermann89) [reply src/mrpro/data/KTrajectory.py:75]: YES!

File: src/mrpro/operators/CartesianSamplingOp.py (hunk 5 above)
Comment (fzimmermann89) [reply src/mrpro/operators/CartesianSamplingOp.py]: remvoed coils here

**Comment by koflera** [comment]
is a 1 missing in the following?

https://github.com/PTB-MR/mrpro/blob/a639f932a1017469bf3b7054da53d2e412192f1f/src/mrpro/data/DcfData.py#L66

**Review by koflera — APPROVED** [review]
looks good to me

**Comment by fzimmermann89** [comment]
> is a 1 missing in the following?
> 
> https://github.com/PTB-MR/mrpro/blob/a639f932a1017469bf3b7054da53d2e412192f1f/src/mrpro/data/DcfData.py#L66

I think it's correct:
Same shape as trajectory, but last three dimensions replaced by 1.

Trajectory got another dimension, so did DCF. So it still matches?
